BaseLLM -> Base language

jacobrosenthal opened the pull request:
This mostly does the chains.  I havent looked into agents yet. I know recent Prs are making a whole new class for agent for the new model anyway..

I looked around the python I dont think ser/de is done there either so not sure of the format or if it needs to be supported in order to merge this.

But tests pass and the touched examples run with either model. I dont know if its time to default examples into the ChatOpenAI class yet?

I have not tested the langchain/base_language export as my typescript tooling is poor these days, still working that out but this would seem to do it.

hwchase17 commented on langchain/src/chains/prompt_selector.ts:
```diff
@@ -38,8 +37,8 @@ export class ConditionalPromptSelector extends BasePromptSelector {
 
-export function isLLM(llm: BaseLanguageModel): llm is BaseLLM {
-  return llm instanceof BaseLLM;
+export function isLLM(llm: BaseLanguageModel): llm is BaseLanguageModel {
```
this should actually stay the same - its checking the subtype

hwchase17 commented on langchain/src/chains/tests/combine_docs_chain.test.ts:
```diff
@@ -1,37 +1,55 @@
 
-  class FakeLLM extends BaseLLM {
-    _llmType(): string {
+  class FakeLLM extends BaseLanguageModel {
```
i think its fine for this to remain the same and not change? eg extend BaseLLM

jacobrosenthal commented:
@hwchase17 Addressed those nits. 

Im also sneaking some exports in here as the reason I need this is Im trying to build/extend agents on the new BaseLanguageModel out of tree

related, since I have your attention :) 

custom agents need to be able to hook agents output, much like you can for custom chains  (instead of extending the entire executor to get access to `_call`) https://github.com/hwchase17/langchainjs/pull/309

nfcampos commented on langchain/src/chains/llm_chain.ts:
```diff
@@ -97,15 +98,15 @@ export class LLMChain extends BaseChain implements LLMChainInput {
 
     return new LLMChain({
-      llm: await BaseLLM.deserialize(serializedLLM),
+      llm: await BaseLanguageModel.deserialize(serializedLLM),
```
This method throws a "not implemented" error, therefore this will fail

jacobrosenthal replied:
That seems like an ok solution, since its not implemented right? Other options?

jacobrosenthal replied:
Probably core team has more view into all the serde stuff. I dont use it atm, and i have no idea what desired standard formats are. Ill have to leave that you all.

nfcampos commented on langchain/src/prompts/index.ts:
```diff
@@ -28,3 +28,4 @@ export {
   SerializedMessagePromptTemplate,
   SerializedChatPromptTemplate,
 } from "./serde.js";
+export { interpolateFString } from "./template.js";
```
This is an implementation detail, which we do not want exported

jacobrosenthal replied:
I often need to create my agents out of tree, which means implementing createPrompt which means people will want to be able to export this or something like it

nfcampos replied:
Yea, for that you should use `parseTemplate` or `renderTemplate` from the same file. I'd be happy to see those exported to everyone.

nfcampos reviewed (commented):
Looks good overall, needs 2 changes

jacobrosenthal commented:
Why doesn't agent use those then. People are going to copy the agent they
want to extend

On Fri, Mar 17, 2023, 10:35 AM Nuno Campos ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In langchain/src/prompts/index.ts
> <https://github.com/hwchase17/langchainjs/pull/327#discussion_r1140520534>
> :
>
> > @@ -28,3 +28,4 @@ export {
>    SerializedMessagePromptTemplate,
>    SerializedChatPromptTemplate,
>  } from "./serde.js";
> +export { interpolateFString } from "./template.js";
>
> Yea, for that you should use parseTemplate or renderTemplate from the
> same file. I'd be happy to see those exported to everyone.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hwchase17/langchainjs/pull/327#discussion_r1140520534>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADPI5A3XP5BOS5JHLEDX33W4SOHBANCNFSM6AAAAAAV26DELQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nfcampos commented:
That's something we need to fix, I'll create an issue to make agent use parseTemplate, thanks for noticing

jacobrosenthal commented:
Rebased to fix the conflicts and squash this mess down.

jacobrosenthal commented:
Moved serde functions to base like Basellm. Not sure thats the best solution, but it matches upstream

Then add a test something like this
```
test.only("Test deserialize", async () => {
  const model = new OpenAI({ modelName: "text-ada-001" });
  const prompt = new PromptTemplate({
    template: "Print {foo}",
    inputVariables: ["foo"],
  });
  const chain = new LLMChain({ prompt, llm: model });

  const serialized = chain.serialize();
  console.log(serialized)
  const chain2 = await LLMChain.deserialize({ ...serialized });

  const res = await chain2.run("my favorite color");
  console.log({ res });

  // chain === chain2?
});
```
upstream prints this (no llm because wasnt supported)
```
{
  _type: 'llm_chain',
  prompt: {
    _type: 'prompt',
    input_variables: [ 'foo' ],
    output_parser: undefined,
    template: 'Print {foo}',
    template_format: 'f-string'
  }
}
```
this branch now prints
```
    {
      _type: 'llm_chain',
      llm: {
        model_name: 'text-ada-001',
        model: 'text-ada-001',
        temperature: 0.7,
        max_tokens: 256,
        top_p: 1,
        frequency_penalty: 0,
        presence_penalty: 0,
        n: 1,
        best_of: 1,
        logit_bias: undefined,
        stop: undefined,
        stream: false,
        apiKey: 'sk-xxxxxxxxxxxxxxxxxxxx',
        _type: 'openai',
        _model: 'base_llm'
      },
      prompt: {
        _type: 'prompt',
        input_variables: [ 'foo' ],
        output_parser: undefined,
        template: 'Print {foo}',
        template_format: 'f-string'
      }
    }
```

Is that enough implementation? Correct fields?
Should llm serialize your apiKey?
other tests I have api keys for still seem to pass, but theres basically no tests for serde
